pkg/upgrade/result/v1alpha1/upgraderesult.go

```diff
@@ -30,8 +33,100 @@ type upgradeResultList struct {
 	items []*upgradeResult
 }
 
-// ListBuilder enables building
-// an instance of urList
+// builder enables building an instance of
+// upgradeResult
+type builder struct {
+	*upgradeResult
+	checks []Predicate
+	errors []error
+}
+
+// Predicate abstracts conditional logic w.r.t the upgradeResult instance
+//
+// NOTE:
+// Predicate is a functional approach versus traditional approach to mix
+// conditions such as *if-else* within blocks of business logic
+//
+// NOTE:
+// Predicate approach enables clear separation of conditionals from
+// imperatives i.e. actions that form the business logic
+type Predicate func(*upgradeResult) (message string, ok bool)
+
+// predicateFailedError returns the provided predicate as an error
+func predicateFailedError(message string) error {
+	return errors.Errorf("predicatefailed: %s", message)
+}
+
+// Builder returns a new instance of builder
+func Builder() *builder {
+	return &builder{upgradeResult: &upgradeResult{
+		object: &apis.UpgradeResult{},
+	}}
+}
+
+// BuilderForRuntask returns a new instance
+// of builder for runtasks
+func BuilderForRuntask(context, yml string, values map[string]interface{}) *builder {
```

Comment:
can we rename arguments as 
`context, templateYaml string, templateValues map[string]interface{}`

Reply:
Done